README.md
# fortios-syslog

### Overview
This repository is a parser for Fortinet Syslogs. 

Data Log
```sh
 { date :: {-# UNPACK #-} !Date
  , time :: {-# UNPACK #-} !TimeOfDay
  , deviceName :: {-# UNPACK #-} !Bytes
  , deviceId :: {-# UNPACK #-} !Bytes
  , logId :: {-# UNPACK #-} !Bytes
  , type_ :: {-# UNPACK #-} !Bytes
  , subtype :: {-# UNPACK #-} !Bytes
  , fields :: !(Chunks Field)
```


Data Attributes
```sh
= Action {-# UNPACK #-} !Bytes
  | Alert {-# UNPACK #-} !Word64
  | App {-# UNPACK #-} !Bytes
  | ApplicationAction {-# UNPACK #-} !Bytes
  | ApplicationCategory {-# UNPACK #-} !Bytes
  | ApplicationId {-# UNPACK #-} !Word64
    -- ^ ID of the application.
  | ApplicationList {-# UNPACK #-} !Bytes
  | ApplicationRisk {-# UNPACK #-} !Bytes
  | Attack {-# UNPACK #-} !Bytes
  | AttackId {-# UNPACK #-} !Word64
    -- ^ Risk level of the application.
  | AuthServer {-# UNPACK #-} !Bytes
  | Category {-# UNPACK #-} !Word64
  | CategoryDescription {-# UNPACK #-} !Bytes
  | CentralNatId {-# UNPACK #-} !Word64
  | CountIps {-# UNPACK #-} !Word64
    -- ^ Number of the IPS logs associated with the session
  | ClientReputationScore {-# UNPACK #-} !Word64
  | ClientReputationLevel {-# UNPACK #-} !Bytes
  | ClientReputationAction {-# UNPACK #-} !Bytes
  | CountApplication {-# UNPACK #-} !Word64
    -- ^ Number of App Ctrl logs associated with the session.
  | CountWeb {-# UNPACK #-} !Word64
  | Description {-# UNPACK #-} !Bytes
  | DestinationCountry {-# UNPACK #-} !Bytes
  | DestinationDeviceCategory {-# UNPACK #-} !Bytes
  | DestinationDeviceType {-# UNPACK #-} !Bytes
  | DestinationHost {-# UNPACK #-} !Bytes
  | DestinationInterface {-# UNPACK #-} !Bytes
  | DestinationInterfaceRole {-# UNPACK #-} !Bytes
  | DestinationInternetService {-# UNPACK #-} !Bytes
  | DestinationIp {-# UNPACK #-} !IP
  | DestinationMac {-# UNPACK #-} !Net.Types.Mac
  | DestinationOsName {-# UNPACK #-} !Bytes
  | DestinationOsVersion {-# UNPACK #-} !Bytes
  | DestinationPort {-# UNPACK #-} !Word16
  | DestinationServer {-# UNPACK #-} !Word64
  | DestinationUuid {-# UNPACK #-} !Word128
  | DeviceCategory {-# UNPACK #-} !Bytes
  | DeviceType {-# UNPACK #-} !Bytes
  | DhcpMessage {-# UNPACK #-} !Bytes
  | Direction {-# UNPACK #-} !Bytes
  | Duration {-# UNPACK #-} !Word64
  | Error {-# UNPACK #-} !Bytes
  | EventTime {-# UNPACK #-} !Word64
  | EventType {-# UNPACK #-} !Bytes
  | Group {-# UNPACK #-} !Bytes
  | Hostname {-# UNPACK #-} !Bytes
  | IncidentSerialNumber {-# UNPACK #-} !Word64
  | Interface {-# UNPACK #-} !Bytes
  | Ip {-# UNPACK #-} !IP
  | LanIn {-# UNPACK #-} !Word64
  | LanOut {-# UNPACK #-} !Word64
  | Lease {-# UNPACK #-} !Word64
  | Level {-# UNPACK #-} !Bytes
  | LocalPort {-# UNPACK #-} !Word16
  | LogDescription {-# UNPACK #-} !Bytes
  | Mac {-# UNPACK #-} !Net.Types.Mac
  | MasterDestinationMac {-# UNPACK #-} !Net.Types.Mac
  | MasterSourceMac {-# UNPACK #-} !Net.Types.Mac
  | Message {-# UNPACK #-} !Bytes
  | Method {-# UNPACK #-} !Bytes
  | NextStatistics {-# UNPACK #-} !Word64
  | OsName {-# UNPACK #-} !Bytes
  | OsVersion {-# UNPACK #-} !Bytes
  | PolicyId {-# UNPACK #-} !Word64
  | PolicyType {-# UNPACK #-} !Bytes
  | PolicyUuid {-# UNPACK #-} !Word128
  | Profile {-# UNPACK #-} !Bytes
  | ProfileType {-# UNPACK #-} !Bytes
  | Protocol {-# UNPACK #-} !Word8
  | QueryClass {-# UNPACK #-} !Bytes
  | QueryName {-# UNPACK #-} !Bytes
  | QueryType {-# UNPACK #-} !Bytes
  | QueryTypeValue {-# UNPACK #-} !Word64
    -- ^ IANA Internet Protocol Number.
  | ReceivedBytes {-# UNPACK #-} !Word64
    -- ^ Number of bytes received.
  | ReceivedDelta {-# UNPACK #-} !Word64
  | ReceivedPackets {-# UNPACK #-} !Word64
    -- ^ Number of packets received.
  | Reference {-# UNPACK #-} !Bytes
  | ReferralUrl {-# UNPACK #-} !Bytes
  | RemoteIp {-# UNPACK #-} !IP
  | RemotePort {-# UNPACK #-} !Word16
  | RequestType {-# UNPACK #-} !Bytes
  | SentBytes {-# UNPACK #-} !Word64
    -- ^ Number of bytes sent.
  | SentDelta {-# UNPACK #-} !Word64
  | SentPackets {-# UNPACK #-} !Word64
    -- ^ Number of packets sent.
  | Service {-# UNPACK #-} !Bytes
    -- ^ Name of the service.
  | SessionId {-# UNPACK #-} !Word64
  | Severity {-# UNPACK #-} !Bytes
  | SourceCountry {-# UNPACK #-} !Bytes
  | SourceFamily {-# UNPACK #-} !Bytes
  | SourceHardwareVendor {-# UNPACK #-} !Bytes
  | SourceInterface {-# UNPACK #-} !Bytes
  | SourceInterfaceRole {-# UNPACK #-} !Bytes
  | SourceIp {-# UNPACK #-} !IP
  | SourceMac {-# UNPACK #-} !Net.Types.Mac
  | SourceName {-# UNPACK #-} !Bytes
  | SourcePort {-# UNPACK #-} !Word16
  | SourceServer {-# UNPACK #-} !Word64
  | SourceSoftwareVersion {-# UNPACK #-} !Bytes
  | SourceUuid {-# UNPACK #-} !Word128
  | SslAction {-# UNPACK #-} !Bytes
  | SslCertificateCommonName {-# UNPACK #-} !Bytes
  | SslCertificateIssuer {-# UNPACK #-} !Bytes
  | TimeZone {-# UNPACK #-} !Int -- ^ Offset from UTC in minutes
  | TransactionId {-# UNPACK #-} !Word64 -- ^ Field is named @xid@.
  | TranslatedSource {-# UNPACK #-} !IPv4 {-# UNPACK #-} !Word16 -- ^ When @trandisp@ is @snat@
  | TranslatedDestination {-# UNPACK #-} !IPv4 {-# UNPACK #-} !Word16 -- ^ When @trandisp@ is @snat@
  | TunnelId {-# UNPACK #-} !Word64
  | TunnelIp {-# UNPACK #-} !IP
  | TunnelType {-# UNPACK #-} !Bytes
  | UnauthenticatedUser {-# UNPACK #-} !Bytes
  | UnauthenticatedUserSource {-# UNPACK #-} !Bytes
  | UtmAction {-# UNPACK #-} !Bytes
  | Url {-# UNPACK #-} !Bytes
  | UrlSource {-# UNPACK #-} !Bytes
  | UrlFilterIndex {-# UNPACK #-} !Word64
  | UrlFilterList {-# UNPACK #-} !Bytes
  | User {-# UNPACK #-} !Bytes
  | VirtualDomain {-# UNPACK #-} !Bytes
  | VirtualWanLinkId {-# UNPACK #-} !Word64
  | VirtualWanLinkQuality {-# UNPACK #-} !Bytes
  | Vpn {-# UNPACK #-} !Bytes
  | VpnType {-# UNPACK #-} !Bytes
  | WanIn {-# UNPACK #-} !Word64
  | WanOut {-# UNPACK #-} !Word64
```

CHANGELOG.md
# Revision history for fortigate-syslog

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Main.hs
{-# language LambdaCase #-}

import Control.Monad (when)
import Data.Bytes (Bytes)
import Data.Foldable (for_)

import qualified Data.Bytes as Bytes
import qualified Fortios.Syslog as FGT
import qualified Sample as S
import qualified Net.IP as IP
import qualified Net.IPv4 as IPv4

main :: IO ()
main = do
  putStrLn "Start"
  putStrLn "traffic_local_A"
  testTrafficLocalA
  putStrLn "traffic_forward_A"
  testTrafficForwardA
  putStrLn "traffic_forward_B"
  testTrafficForwardB
  putStrLn "utm_webfilter_A"
  testUtmWebfilterA
  putStrLn "utm_webfilter_B"
  testUtmWebfilterB
  putStrLn "utm_webfilter_C"
  testUtmWebfilterC
  putStrLn "event_system_A"
  testEventSystemA
  pure ()

testTrafficLocalA :: IO ()
testTrafficLocalA = case FGT.decode S.traffic_local_A of
  Left e -> fail (show e)
  Right x -> do
    when (FGT.type_ x /= str "traffic")
      (fail "wrong type")
    when (FGT.deviceId x /= str "FGT-Device")
      (fail "wrong device id")
    for_ (FGT.fields x)
      (\case
        FGT.ClientReputationLevel v ->
          when (v /= str "high") (fail "wrong crlevel")
        FGT.SessionId v ->
          when (v /= 69597381) (fail "wrong sessionid")
        FGT.DestinationCountry v ->
          when (v /= str "United States") (fail "wrong dstcountry")
        FGT.SourceIp v ->
          when (v /= IP.ipv4 192 0 2 12) (fail "wrong srcip")
        _ -> pure ()
      ) 

testTrafficForwardA :: IO ()
testTrafficForwardA = case FGT.decode S.traffic_forward_A of
  Left e -> fail (show e)
  Right x -> do
    when (FGT.subtype x /= str "forward")
      (fail "wrong subtype")
    for_ (FGT.fields x)
      (\case
        FGT.CountWeb v ->
          when (v /= 1) (fail "wrong countweb")
        FGT.TranslatedSource ip port -> do
          when (ip /= IPv4.fromOctets 192 0 2 3) (fail "wrong transip")
          when (port /= 54400) (fail "wrong transport")
        _ -> pure ()
      ) 

testTrafficForwardB :: IO ()
testTrafficForwardB = case FGT.decode S.traffic_forward_B of
  Left e -> fail (show e)
  Right x -> do
    when (FGT.subtype x /= str "forward")
      (fail "wrong subtype")
    for_ (FGT.fields x)
      (\case
        FGT.TranslatedDestination ip port -> do
          when (ip /= IPv4.fromOctets 192 0 2 200) (fail "wrong tranip")
          when (port /= 443) (fail "wrong tranport")
        _ -> pure ()
      ) 

testUtmWebfilterA :: IO ()
testUtmWebfilterA = case FGT.decode S.utm_webfilter_A of
  Left e -> fail (show e)
  Right x -> do
    when (FGT.subtype x /= str "webfilter") (fail "wrong subtype")
    for_ (FGT.fields x)
      (\case
        FGT.Profile name -> when (name /= str "my-profile") (fail "profile")
        _ -> pure ()
      ) 

testUtmWebfilterB :: IO ()
testUtmWebfilterB = case FGT.decode S.utm_webfilter_B of
  Left e -> fail (show e)
  Right x -> do
    when (FGT.subtype x /= str "webfilter") (fail "wrong subtype")
    for_ (FGT.fields x)
      (\case
        FGT.Profile name -> when (name /= str "default") (fail "profile")
        FGT.Error name -> when (name /= str "DNS query timeout") (fail "error")
        _ -> pure ()
      )

testUtmWebfilterC :: IO ()
testUtmWebfilterC = case FGT.decode S.utm_webfilter_C of
  Left e -> fail (show e)
  Right x -> do
    when (FGT.subtype x /= str "webfilter") (fail "wrong subtype")
    for_ (FGT.fields x)
      (\case
        FGT.Profile name -> when (name /= str "prof") (fail "profile")
        FGT.ReferralUrl name -> when
          (name /= str "http://www.example.org/get/started")
          (fail "referralurl")
        _ -> pure ()
      )

testEventSystemA :: IO ()
testEventSystemA = case FGT.decode S.event_system_A of
  Left e -> fail (show e)
  Right x -> do
    when (FGT.subtype x /= str "system") (fail "wrong subtype")
    for_ (FGT.fields x)
      (\case
        FGT.Profile name -> when (name /= str "prof") (fail "profile")
        FGT.DhcpMessage name -> when (name /= str "Ack") (fail "dhcp_message")
        FGT.Interface name ->
          when (name /= str "vlan-11-trust") (fail "interface")
        FGT.LogDescription name ->
          when (name /= str "DHCP Ack log") (fail "logdesc")
        FGT.Lease n -> when (n /= 3600) (fail "lease")
        _ -> pure ()
      )

str :: String -> Bytes
str = Bytes.fromAsciiString


common/Sample.hs
{-# language TypeApplications #-}

module Sample
  ( traffic_local_A
  , traffic_forward_A
  , traffic_forward_B
  , utm_webfilter_A
  , utm_webfilter_B
  , utm_webfilter_C
  , event_system_A
  ) where

import Data.Bytes (Bytes)
import Data.Word (Word8)
import Data.Char (ord)
import qualified Data.Bytes as Bytes
import qualified GHC.Exts as Exts

-- Sample Logs. If you add a sample log to this file, please
-- replace all information in the log that could possibly be
-- meaningful. At a bare minimum, this means:
--
-- * Replace any IP addresses with non-routable addresses 
--   from the TEST-NET-1 block (192.0.2.0/24).
-- * Replace any domain names with the reserved domain
--   name example.com.
-- * Replace any hostnames with something like MY-HOST
--   or NY-APP or SAMPLE-HOST.
-- * Replace rule names and application categories.

pack :: String -> Bytes
pack = Bytes.fromByteArray . Exts.fromList . map (fromIntegral @Int @Word8 . ord)

-- Local Traffic Log 
traffic_local_A :: Bytes
traffic_local_A = pack $ concat
  [ "date=2019-11-08 time=09:50:39 devname=FGT-Device-Node0 "
  , "devid=FGT-Device logid=0001000014 type=traffic "
  , "subtype=local level=notice vd=root srcip=192.0.2.12 "
  , "srcport=48216 srcintf=\"port7\" dstip=192.0.2.23 dstport=14560 "
  , "dstintf=\"root\" sessionid=69597381 proto=6 action=deny policyid=0 "
  , "policytype=some-policy dstcountry=\"United States\" "
  , "srccountry=\"Canada\" trandisp=noop service=\"tcp/14560\" "
  , "app=\"tcp/14560\" duration=0 sentbyte=0 rcvdbyte=0 sentpkt=0 "
  , "appcat=\"ex-cat\" crscore=30 craction=193181 crlevel=high"
  ]

traffic_forward_A :: Bytes
traffic_forward_A = pack $ concat
  [ "date=2019-11-08 time=09:50:39 devname=FGT5-NY-Node0 devid=FGT5-NY "
  , "logid=0000000013 type=traffic subtype=forward level=notice vd=root "
  , "srcip=192.0.2.55 srcport=54400 srcintf=\"vlan-13-trust\" "
  , "dstip=192.0.2.56 dstport=80 dstintf=\"port55\" "
  , "poluuid=176a120e-56c3-51a6-5fb1-485f3180eb81 sessionid=79091892 "
  , "proto=6 action=close policyid=6 policytype=policy "
  , "dstcountry=\"United States\" srccountry=\"Reserved\" trandisp=snat "
  , "transip=192.0.2.3 transport=54400 service=\"HTTP\" duration=6 "
  , "sentbyte=395 rcvdbyte=651 sentpkt=6 rcvdpkt=3 appcat=\"unscanned\" "
  , "wanin=372 wanout=131 lanin=145 lanout=381 utmaction=allow countweb=1"
  ]

traffic_forward_B :: Bytes
traffic_forward_B = pack $ concat
  [ "date=2019-11-15 time=08:24:34 devname=FGT-NY-Node0 devid=FGT-NY "
  , "logid=0000000013 type=traffic subtype=forward level=notice vd=root "
  , "srcip=192.0.2.103 srcport=55397 srcintf=\"port2\" dstip=192.0.2.97 "
  , "dstport=443 dstintf=\"vlan-1-trust\" poluuid=ecf7d054-16d2-50e7-e0c1-"
  , "fb0529e4ab90 sessionid=97182243 proto=6 action=close policyid=22 "
  , "policytype=policy dstcountry=\"United States\" srccountry=\"United States\" "
  , "trandisp=dnat tranip=192.0.2.200 tranport=443 service=\"HTTPS\" duration=97 "
  , "sentbyte=1204 rcvdbyte=935 sentpkt=9 rcvdpkt=6 appcat=\"unscanned\""
  ]

utm_webfilter_A :: Bytes
utm_webfilter_A = pack $ concat
  [ "date=2019-11-15 time=08:24:33 devname=FGT-NY-Node0 devid=FGT-NY "
  , "logid=0317013312 type=utm subtype=webfilter eventtype=ftgd_allow "
  , "level=notice vd=root policyid=6 sessionid=97184671 user=\"\" "
  , "srcip=192.0.2.31 srcport=53746 srcintf=\"vlan-7-trust\" "
  , "dstip=192.0.2.51 dstport=80 dstintf=\"port5\" proto=6 service=\"HTTP\" "
  , "hostname=\"192.0.2.201\" profile=\"my-profile\" action=passthrough "
  , "reqtype=direct url=\"/example/path.txt\" sentbyte=93 rcvdbyte=0 "
  , "direction=outgoing msg=\"URL belongs to an allowed category in policy\" "
  , "method=domain cat=52 catdesc=\"Information Technology\""
  ]

utm_webfilter_B :: Bytes
utm_webfilter_B = pack $ concat
  [ "date=2019-11-18 time=14:11:36 devname=FGT-NY-Node0 devid=FGT-NY "
  , "logid=0318012800 type=utm subtype=webfilter eventtype=ftgd_err "
  , "level=error vd=root policyid=5 sessionid=106167548 user=\"\" "
  , "srcip=192.0.2.67 srcport=58737 srcintf=\"vlan-115\" dstip=192.0.2.100 "
  , "dstport=53 dstintf=\"port59\" proto=17 service=\"DNS\" "
  , "hostname=\"example.com\" profile=\"default\" action=blocked "
  , "reqtype=direct sentbyte=0 rcvdbyte=0 direction=outgoing "
  , "msg=\"A rating error occurs\" error=\"DNS query timeout\""
  ]

utm_webfilter_C :: Bytes
utm_webfilter_C = pack $ concat
  [ "date=2019-11-18 time=14:11:35 devname=FGT-NY-Node0 devid=FGT-NY "
  , "logid=0317013312 type=utm subtype=webfilter eventtype=ftgd_allow "
  , "level=notice vd=root policyid=6 sessionid=107172027 user=\"\" "
  , "srcip=192.0.2.3 srcport=52992 srcintf=\"vlan-19-trust\" "
  , "dstip=192.0.2.54 dstport=80 dstintf=\"port91\" proto=6 "
  , "service=\"HTTP\" hostname=\"www.example.com\" "
  , "profile=\"prof\" action=passthrough reqtype=referral "
  , "url=\"/the/path.php?name=yes\" "
  , "referralurl=\"http://www.example.org/get/started\" "
  , "sentbyte=38053 rcvdbyte=126033 direction=outgoing "
  , "msg=\"URL belongs to an allowed category in policy\" "
  , "method=domain cat=42 catdesc=\"Shopping\""
  ]

event_system_A :: Bytes
event_system_A = pack $ concat
  [ "date=2019-11-18 time=14:44:07 devname=FGT-NY-Node0 devid=FGT-NY "
  , "logid=0100026001 type=event subtype=system level=information vd=root "
  , "logdesc=\"DHCP Ack log\" interface=\"vlan-11-trust\" dhcp_msg=\"Ack\" "
  , "mac=D1:1F:B1:41:FB:1F ip=192.0.2.97 lease=3600 hostname=\"SomeHost\" "
  , "msg=\"DHCP server sends a DHCPACK\""
  ]


src/Fortios/Hash.hs
{-# language BangPatterns #-}
{-# language TypeApplications #-}

module Fortios.Hash
  ( duohash
  , quadrohash
  ) where

import Data.Bytes.Types (Bytes(Bytes))
import Data.Primitive (indexByteArray)
import Data.Word (Word8)

duohash :: Int -> Int -> Int -> Bytes -> Word
duohash !acc !x1 !x2 (Bytes !arr !off !len) = case len of
  0 -> fromIntegral @Int @Word acc
  _ ->
    let accNew = acc
          + x1 * fromIntegral (indexByteArray arr off :: Word8)
          + x2 * fromIntegral (indexByteArray arr (off + 1) :: Word8)
     in duohash accNew x1 x2 (Bytes arr (off + 2) (len - 2)) 
  
quadrohash :: Int -> Int -> Int -> Int -> Int -> Bytes -> Word
quadrohash !acc !x1 !x2 !x3 !x4 (Bytes !arr !off !len) = case len of
  0 -> fromIntegral @Int @Word acc
  _ ->
    let accNew = acc
          + x1 * fromIntegral (indexByteArray arr off :: Word8)
          + x2 * fromIntegral (indexByteArray arr (off + 1) :: Word8)
          + x3 * fromIntegral (indexByteArray arr (off + 2) :: Word8)
          + x4 * fromIntegral (indexByteArray arr (off + 3) :: Word8)
     in quadrohash accNew x1 x2 x3 x4 (Bytes arr (off + 4) (len - 4)) 


scripts/Generate.hs
{-# language BangPatterns #-}
{-# language LambdaCase #-}
{-# language TypeApplications #-}
{-# language ScopedTypeVariables #-}

-- Generate the Fortios.Generated module. This creates a perfect
-- hash function for strings of each length. That is, there is a
-- separate hash function for strings of length 2, 3, 4, etc.
--
-- The module this produces includes both the hash functions
-- (about 15 of them) and pattern synonyms for the computed hash
-- of each known key (about 50 of them).
module Main where

import Data.Int (Int32)
import Data.Word (Word8,Word16)
import Data.Char (ord)
import Data.Map (Map)
import System.Random (randomIO)
import Control.Monad (replicateM)
import System.IO (stdin)

import qualified Data.Map.Strict as Map
import qualified Data.List as L
import qualified Data.Text as T
import qualified Data.Text.IO as T
import qualified System.IO as IO

-- Intentionally omitted:
--
-- * referralurl: only key of length 11
-- * transip and transport: always follow trandisp=snat
-- * tranip and tranport: always follow trandisp=dnat
allKeywords :: [String]
allKeywords =
  [ "action"
  , "app"
  , "appcat"
  , "cat"
  , "catdesc"
  , "countweb"
  , "craction"
  , "crlevel"
  , "crscore"
  , "date"
  , "devid"
  , "devname"
  , "dhcp_msg"
  , "direction"
  , "dstcountry"
  , "dstintf"
  , "dstip"
  , "dstport"
  , "duration"
  , "error"
  , "eventtype"
  , "hostname"
  , "interface"
  , "ip"
  , "lanin"
  , "lanout"
  , "lease"
  , "level"
  , "logdesc"
  , "logid"
  , "mac"
  , "method"
  , "msg"
  , "policyid"
  , "policytype"
  , "poluuid"
  , "profile"
  , "proto"
  , "rcvdbyte"
  , "rcvdpkt"
  , "reqtype"
  , "sentbyte"
  , "sentpkt"
  , "service"
  , "sessionid"
  , "srccountry"
  , "srcintf"
  , "srcip"
  , "srcport"
  , "subtype"
  , "time"
  , "trandisp"
  , "type"
  , "url"
  , "user"
  , "utmaction"
  , "wanin"
  , "wanout"
  , "vd"
  ]

data Algorithm = AlgoTwo !Int !Int | AlgoFour !Int !Int !Int !Int
  deriving (Show)

type Result = Map Int (Algorithm, Int, Map Word String)

keywordsOfLength :: Int -> [String]
keywordsOfLength !n =
  filter (\k -> length k == n) allKeywords

groupedKeywords :: [[String]]
groupedKeywords = map keywordsOfLength [0..20]

main :: IO ()
main = do
  let go !acc !_ [] = pure acc
      go !acc !len (keywords : xs) = do
        r <- attemptTableSize keywords len
        go (acc <> r) (len + 1) xs
  res <- go mempty 0 groupedKeywords
  Map.foldMapWithKey (\k v -> putStrLn (show k ++ ": " ++ show v)) res
  IO.withFile "src/Fortios/Generated.hs" IO.WriteMode $ \h -> do
    IO.hPutStrLn h "{-# language PatternSynonyms #-}"
    IO.hPutStr h "\n--Notice: Generated by scripts/Generate.hs\n"
    IO.hPutStr h "--This module is generated. Do not modify its contents by hand.\n"
    IO.hPutStr h "module Fortios.Generated\n  ("
    IO.hPutStrLn h (drop 3 (makePatternExports res))
    IO.hPutStrLn h (exportHashFuncs res)
    IO.hPutStrLn h "  ) where\n"
    IO.hPutStrLn h "import Fortios.Hash (duohash,quadrohash)"
    IO.hPutStrLn h "import Data.Bytes.Types (Bytes(Bytes))\n"
    IO.hPutStrLn h "import Data.Primitive (ByteArray)\n"
    IO.hPutStrLn h (makePatterns res)
    IO.hPutStrLn h (makeHashFuncs res)

takeDividingTwo :: [a] -> [a]
takeDividingTwo (x : y : zs) = x : y : takeDividingTwo zs
takeDividingTwo (_ : []) = []
takeDividingTwo [] = []

takeDividingFour :: [a] -> [a]
takeDividingFour (w: x : y : z : zs) = w : x : y : z : takeDividingFour zs
takeDividingFour (_ : _ : _ : []) = []
takeDividingFour (_ : _ : []) = []
takeDividingFour (_ : []) = []
takeDividingFour [] = []

attemptTableSize :: [String] -> Int -> IO Result
attemptTableSize keywords !len = if length keywords > 0  
  then do
    putStrLn ("Mod " ++ show sz ++ " hashes for strings of length " ++ show len ++ ":\n" ++ show keywords)
    attemptMultiplierSizeTwo keywords sz >>= \case
      Just res -> pure (Map.singleton len res)
      Nothing -> attemptMultiplierSizeFour keywords sz >>= \case
        Just res -> pure (Map.singleton len res)
        Nothing -> attemptMultiplierSizeTwo keywords (sz * 2) >>= \case
          Nothing -> fail "Could not come up with hash function"
          Just res -> pure (Map.singleton len res)
  else pure Map.empty
  where
  sz = nextPowerOfTwo (length keywords)

attemptMultiplierSizeTwo :: [String] -> Int
  -> IO (Maybe (Algorithm, Int, Map Word String))
attemptMultiplierSizeTwo keywords !sz =
  let go !i !cutoff = if i >= (0 :: Int)
        then do
          ms0 :: [Word16] <- replicateM 2 randomIO
          -- putStrLn ("Attempt mutipliers: " ++ show ms0)
          let ms1@[m0,m1] = map (fromIntegral @Word16 @Int) ms0
          let ms = streamFromList ms1
          let rs = L.nub (map (hashStringWith sz ms . takeDividingTwo) keywords)
          if L.length rs == L.length keywords && maximum rs <= fromIntegral @Int @Word cutoff
            then do
              putStrLn "Found a match"
              putStrLn ("Multipliers: " ++ show ms1)
              putStrLn ("Hashes: " ++ show rs)
              putStrLn ("Table Size: " ++ show sz)
              let algo = AlgoTwo m0 m1
              let xs = map (\x -> (hashStringWith sz ms (takeDividingTwo x), x)) keywords
              pure (Just (algo,sz,Map.fromList xs))
            else go (i - 1) cutoff
        else if cutoff < sz
          then go 80000 (cutoff + 1)
          else pure Nothing
   in go 80000 (length keywords)

attemptMultiplierSizeFour :: [String] -> Int
  -> IO (Maybe (Algorithm, Int, Map Word String))
attemptMultiplierSizeFour keywords !sz =
  let go !i !cutoff = if i >= (0 :: Int)
        then do
          ms0 :: [Word16] <- replicateM 4 randomIO
          let ms1@[m0,m1,m2,m3] = map (fromIntegral @Word16 @Int) ms0
          let ms = streamFromList ms1
          let rs = L.nub (map (hashStringWith sz ms . takeDividingFour) keywords)
          if L.length rs == L.length keywords && maximum rs <= fromIntegral @Int @Word cutoff
            then do
              putStrLn "Found a match"
              putStrLn ("Multipliers: " ++ show ms1)
              putStrLn ("Hashes: " ++ show rs)
              putStrLn ("Table Size: " ++ show sz)
              let algo = AlgoFour m0 m1 m2 m3
              let xs = map (\x -> (hashStringWith sz ms (takeDividingFour x), x)) keywords
              pure (Just (algo,sz,Map.fromList xs))
            else go (i - 1) cutoff
        else if cutoff < sz
          then go 80000 (cutoff + 1)
          else pure Nothing
   in go 80000 (length keywords)
       
  
streamFromList :: [Int] -> IntStream
streamFromList xs = go xs where
  go [] = go xs
  go (y : ys) = IntStream y (go ys)

data IntStream = IntStream !Int IntStream

hashStringWith :: Int -> IntStream -> String -> Word
hashStringWith !tblSz = go 0 where
  go !acc _ [] = mod (fromIntegral @Int @Word acc) (fromIntegral @Int @Word tblSz)
  go !acc (IntStream m ms) (c : cs) = go (acc + ord c * m) ms cs
  
nextPowerOfTwo :: Int -> Int
nextPowerOfTwo !n = go 1 where
  go !x = if x >= n
    then x
    else go (x * 2)

makePatternExports :: Map Int (Algorithm, Int, Map Word String) -> String
makePatternExports = concat . L.sort . foldMap
  (\(_,_,m) -> Map.foldMapWithKey
    ( \theHash str -> ["  , pattern H_" ++ str ++ "\n"]
    ) m
  )

makePatterns :: Map Int (Algorithm, Int, Map Word String) -> String
makePatterns = concat . L.sort . foldMap
  (\(_,_,m) -> Map.foldMapWithKey
    ( \theHash str ->
      [ "pattern H_" ++ str ++ " :: Word\n"
      , "pattern H_" ++ str ++ " = " ++ show theHash ++ "\n\n"
      ]
    ) m
  )

makeHashFuncs :: Map Int (Algorithm, Int, Map Word String) -> String
makeHashFuncs = Map.foldMapWithKey
  (\len (algo,n,_) -> concat
    [ "hashString" ++ show len ++ " :: ByteArray -> Int -> Word\n"
    , "hashString" ++ show len ++ " arr off = "
    , case algo of
        AlgoTwo x1 x2 ->
          "rem (duohash 0 " ++ show x1 ++ " " ++ show x2 ++ 
          " (Bytes arr off " ++ show (2 * div len 2) ++ ")) " ++
          show n ++ "\n\n"
        AlgoFour x1 x2 x3 x4 ->
          "rem (quadrohash 0 " ++ show x1 ++ " " ++ show x2 ++ " " ++
          show x3 ++ " " ++ show x4 ++ 
          " (Bytes arr off " ++ show (4 * div len 4) ++ ")) " ++
          show n ++ "\n\n"
    ]
  )

exportHashFuncs :: Map Int (Algorithm, Int, Map Word String) -> String
exportHashFuncs = Map.foldMapWithKey
  (\len _ -> "  , hashString" ++ show len ++ "\n"
  )
